src/app/filelogger.cpp

```diff
@@ -0,0 +1,179 @@
+    m_flusher.setInterval(0);
+    m_flusher.setSingleShot(true);
+    connect(&m_flusher, SIGNAL(timeout()), SLOT(flushLog()));
+
+    changePath(path);
+    if (deleteOld)
+        this->deleteOld(age, ageType);
+
+    const Logger* const logger = Logger::instance();
+    foreach (const Log::Msg& msg, logger->getMessages())
+        addLogMessage(msg);
+
+    connect(logger, SIGNAL(newLogMessage(const Log::Msg &)), SLOT(addLogMessage(const Log::Msg &)));
+}
+
+FileLogger::~FileLogger()
+{
+    if (!m_logFile) return;
+    closeLogFile();
+    delete m_logFile;
+}
+
+void FileLogger::changePath(const QString& newPath)
+{
+    QString tmpPath = Utils::Fs::fromNativePath(newPath);
+    if (!tmpPath.endsWith('/'))
+        tmpPath += "/";
+    QDir dir(tmpPath);
+    dir.mkpath(tmpPath);
+    tmpPath += "qbittorrent.log";
+
+    if (tmpPath != m_path) {
+        m_path = tmpPath;
+
+        if (m_logFile) {
+            closeLogFile();
+            delete m_logFile;
+        }
+        m_logFile = new QFile(m_path);
+        openLogFile();
+    }
+}
+
+void FileLogger::deleteOld(const int age, const FileLogAgeType ageType)
+{
+    QDateTime date = QDateTime::currentDateTime();
+    QDir dir(m_path);
+
+    switch (ageType) {
+    case DAYS:
+        date = date.addDays(age);
+        break;
+    case MONTHS:
+        date = date.addMonths(age);
+        break;
+    default:
+        date = date.addYears(age);
+    }
+
+    foreach (const QFileInfo file, dir.entryInfoList(QStringList() << "qbittorrent.log.bak*", QDir::Files | QDir::Writable, QDir::Time | QDir::Reversed)) {
+        if (file.lastModified() < date)
+            break;
+        Utils::Fs::forceRemove(file.absoluteFilePath());
+    }
+}
+
+void FileLogger::setBackup(bool value)
+{
+    m_backup = value;
+}
+
+void FileLogger::setMaxSize(int value)
+{
+    m_maxSize = value;
+}
+
+void FileLogger::addLogMessage(const Log::Msg &msg)
+{
+    if (!m_logFile) return;
+
+    QTextStream str(m_logFile);
+    QString type;
+
+    switch (msg.type) {
+    case Log::INFO:
+        type = "(I) ";
+        break;
+    case Log::WARNING:
+        type = "(W) ";
+        break;
+    case Log::CRITICAL:
+        type = "(C) ";
+        break;
+    default:
+        type = "(N) ";
+    }
+
+    str << type << QDateTime::fromMSecsSinceEpoch(msg.timestamp).toString(Qt::ISODate) << " - " << msg.message << endl;
+
+    if (m_backup && (m_logFile->size() >= (m_maxSize * 1024 * 1024))) {
+        closeLogFile();
+        int counter = 0;
+        QString backupLogFilename = m_path + ".bak";
+
+        while (QFile::exists(backupLogFilename)) {
+            ++counter;
+            backupLogFilename = m_path + ".bak" + QString::number(counter);
+        }
+
+        QFile::rename(m_path, backupLogFilename);
+        openLogFile();
+    }
+    else {
+        m_flusher.start();
+    }
+}
+
+void FileLogger::flushLog()
+{
+    if (m_logFile)
+        m_logFile->flush();
+}
+
+void FileLogger::openLogFile()
+{
```

Comment:
Wouldn't it be easier if you put the file creation here, and deleting in `closeLogFile()`?
I know that in one place it will lead to we create it immediately after removal. For me, it's a small problem.